### PR TITLE
Test SPO vote counting

### LINKED ISSUE #4617: Add threshold property tests for SPO voting
#4326 states that both `Committee` and `DRep` voting have property tests that check the math behind voting thresholds. It might be a good idea to do the same for SPO voting as well.

### LINKED ISSUE #4727: Bring back couple of Imp tests that check SPO voting
The test removed in #4713 was unclear because it evolved from a more clear test into something that expected to be tested in bootstrap and after bootstrap. 

If I remember correctly this test(s) where suppose to check SPOs voting.

Instead of removing tests it would be better to adjust them to make them work in the new re-organized imp spec. I'd suggest one of them to run in the post-bootsrap only, while the other switching to using a different action, like PParamUpdate.

_Originally posted by @lehins in https://github.com/IntersectMBO/cardano-ledger/pull/4713#discussion_r1819632627_

So, this ticket is about bringing those tests back, but with more sensible semantics:
```haskell
spoVotesCommitteeUpdates ::
  forall era.
  ConwayEraImp era =>
  SpecWith (ImpTestState era)
spoVotesCommitteeUpdates =
  describe "Counting of SPO votes" $ do
    describe "All gov action other than HardForkInitiation" $ do
      it "NoConfidence" $ do
        (spoK1, _, _) <- setupPoolWithStake $ Coin 100_000_000
        _ <- setupPoolWithStake $ Coin 100_000_000
        _ <- setupPoolWithStake $ Coin 100_000_000
        _ <- setupPoolWithStake $ Coin 100_000_000
        modifyPParams $ ppPoolVotingThresholdsL . pvtMotionNoConfidenceL .~ 1 %! 2
        whenPostBootstrap (modifyPParams $ ppDRepVotingThresholdsL .~ def)
        gai <- submitGovAction $ NoConfidence SNothing
        -- bootstrap: 1 % 4 stake yes; 3 % 4 stake abstain; yes / stake - abstain > 1 % 2
        -- post-bootstrap: 1 % 4 stake yes; 3 % 4 stake no; yes stake < 1 % 2
        submitYesVote_ (StakePoolVoter spoK1) gai
        passNEpochs 2
        pv <- getProtVer
        if bootstrapPhase pv
          then getLastEnactedCommittee `shouldReturn` SJust (GovPurposeId gai)
          else getLastEnactedCommittee `shouldReturn` SNothing
      it "CommitteeUpdate" $ do
        (spoK1, _, _) <- setupPoolWithStake $ Coin 100_000_000
        _ <- setupPoolWithStake $ Coin 100_000_000
        _ <- setupPoolWithStake $ Coin 100_000_000
        _ <- setupPoolWithStake $ Coin 100_000_000
        modifyPParams $ ppPoolVotingThresholdsL . pvtCommitteeNormalL .~ 1 %! 2
        whenPostBootstrap (modifyPParams $ ppDRepVotingThresholdsL .~ def)
        cc <- KeyHashObj <$> freshKeyHash
        gai <- submitUpdateCommittee Nothing mempty [(cc, EpochInterval 5)] (1 %! 2)
        -- bootstrap: 1 % 4 stake yes; 3 % 4 stake abstain; yes / stake - abstain > 1 % 2
        -- post-bootstrap: 1 % 4 stake yes; 3 % 4 stake no; yes stake < 1 % 2
        submitYesVote_ (StakePoolVoter spoK1) gai
        passNEpochs 2
        pv <- getProtVer
        if bootstrapPhase pv
          then getLastEnactedCommittee `shouldReturn` SJust (GovPurposeId gai)
          else getLastEnactedCommittee `shouldReturn` SNothing
```

### PR DESCRIPTION
**Pull request by Lucsanszky**
# Description
Follow-up for #4659 
Resolves #4617
Resolves #4727

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages.
      **_New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated.
      **_If you change the bounds in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: eras/conway/impl/test/Test/Cardano/Ledger/Conway/SPORatifySpec.hs
Comment (aniketd) [inline eras/conway/impl/test/Test/Cardano/Ledger/Conway/SPORatifySpec.hs]: It is not easy to read that the two calls to `spoAcceptedRatio` are the same. Perhaps making a common reference will help in making it easier to understand.
```suggestion
              let acceptedRatio =
                    spoAcceptedRatio
                      @era 
                      re {reStakePoolDistr = distr}
                      gas {gasStakePoolVotes = votes}
                      (rs ^. rsEnactStateL . ensProtVerL)
               in if fromCompact totalStake == Coin 0
                     then acceptedRatio `shouldBe` 0
                     else acceptedRatio `shouldBe` 1
                
```

File: eras/conway/impl/test/Test/Cardano/Ledger/Conway/SPORatifySpec.hs
Comment (aniketd) [inline eras/conway/impl/test/Test/Cardano/Ledger/Conway/SPORatifySpec.hs]: Why not
```suggestion
  pools <- listOf (arbitrary @(KeyHash 'StakePool (EraCrypto era)))
```

File: eras/conway/impl/test/Test/Cardano/Ledger/Conway/SPORatifySpec.hs
Comment (aniketd) [inline eras/conway/impl/test/Test/Cardano/Ledger/Conway/SPORatifySpec.hs]: At first I didn't understand why one of the values is being ignored. 😃 👍 
```suggestion
-- | Generates 6 rational values for voting ratios. One of them should be ignored to account for stake that didn't participate.
genPctsOf100 :: Gen (Rational, Rational, Rational, Rational, Rational, Rational)
```

**Review by aniketd — COMMENTED** [review]
Just a few minor suggestions. LGTM! 👍

File: eras/conway/impl/test/Test/Cardano/Ledger/Conway/SPORatifySpec.hs
Comment (lehins) [inline eras/conway/impl/test/Test/Cardano/Ledger/Conway/SPORatifySpec.hs]: This simplification uses [`fromKeys` function that we have available](https://github.com/IntersectMBO/cardano-ledger/blob/97c2c87a710d6b122cee6a37f47c0aa952ae7bef/libs/cardano-data/src/Data/MapExtras.hs#L251C1-L251C9) as well as the Monoid instance of `Map`
```suggestion
    unionAllFromLists = foldMap (\(ks, v) -> fromKeys (const v) ks)
```

File: eras/conway/impl/test/Test/Cardano/Ledger/Conway/SPORatifySpec.hs
Comment (lehins) [inline eras/conway/impl/test/Test/Cardano/Ledger/Conway/SPORatifySpec.hs]: This simplification uses [`fromKeys` function that we have available](https://github.com/IntersectMBO/cardano-ledger/blob/97c2c87a710d6b122cee6a37f47c0aa952ae7bef/libs/cardano-data/src/Data/MapExtras.hs#L251C1-L251C9)
```suggestion
      fromKeys (const drep) . map (raCredential . ppRewardAccount) . Map.elems
```

File: eras/conway/impl/test/Test/Cardano/Ledger/Conway/SPORatifySpec.hs
Comment (lehins) [inline eras/conway/impl/test/Test/Cardano/Ledger/Conway/SPORatifySpec.hs]: This simplification uses [`fromKeys` function that we have available](https://github.com/IntersectMBO/cardano-ledger/blob/97c2c87a710d6b122cee6a37f47c0aa952ae7bef/libs/cardano-data/src/Data/MapExtras.hs#L251C1-L251C9)

```suggestion
      pure $ fromKeys (const params) p
```

However, both my suggestion above and the implementation will use the same pool params, which seems wrong to me. Not only because it didn't look like that is what we need in the tests, but also because with this generation the invariant is violated where params contain the same poolId as the key in the Map.

It seems to me this was the intended generator:
```suggestion
      let genPoolParams poolId = do
            poolParams <- arbitrary
            pure $ poolParams { ppId = poolId }
      sequence $ fromKeys genPoolParams p
```

I can't fully understand the tests yet to tell for sure which one of these is the desired generators.

File: eras/conway/impl/test/Test/Cardano/Ledger/Conway/SPORatifySpec.hs
Comment (lehins) [inline eras/conway/impl/test/Test/Cardano/Ledger/Conway/SPORatifySpec.hs]:
```suggestion
      , poolParams = Map.unions [poolParamsRest, poolParamsAA, poolParamsNC]
```

**Comment by Lucsanszky** [comment]
Note to whoever takes this over as I'm off this week: I reinstated the previously removed tests in the last commit. I only adjusted them minimally, so they pass. I think the one with the `NoConfidence` action could be removed as I often used `NoConfidence` in the Imp tests to check the vote counting behaviour. The committee update one could stay or maybe adjusted to better fit the recently added tests. I also wouldn't be against dropping that commit altogether and not reinstating those old tests at all because committee updates are also covered by the newly added `acceptedRatioProp` as that could generate such cases.

File: eras/conway/impl/test/Test/Cardano/Ledger/Conway/SPORatifySpec.hs
Comment (lehins) [inline eras/conway/impl/test/Test/Cardano/Ledger/Conway/SPORatifySpec.hs]: Minor suggestion to improve readability:
```suggestion
        let re' = re {reStakePoolDistr = PoolDistr Map.empty (fromJust . toCompact $ Coin 100)}
             in spoAccepted @era re' rs gas
                  `shouldBe` 
                  (votingStakePoolThreshold @era rs (gasAction gas) == SJust minBound)
```

**Comment by Lucsanszky** [comment]
Regarding #4727, I dropped the `NoConfidence` test since we cover that case quite extensively and I changed one of the newly added tests to cover the `CommitteeUpdate` case. I think this is now good to go, so I'll enable auto-merge and wait for approval.

**Review by lehins — APPROVED** [review]
:+1: 
Thank you!

**Pull request merged**
